[RISCV] Move vmv.v.v peephole from SelectionDAG to RISCVVectorPeephole

lukel97 opened the pull request:
This is split off from #71764, and moves only the vmv.v.v part of performCombineVMergeAndVOps to work on MachineInstrs.

In retrospect trying to handle PseudoVMV_V_V and PseudoVMERGE_VVM in the same function makes the code quite hard to read, so this just does it in a separate peephole.

This turns out to be simpler since for PseudoVMV_V_V we don't need to convert the Src instruction to a masked variant, and we don't need to create a fake all ones mask.

lukel97 committed:
[RISCV] Move vmv.v.v peephole from SelectionDAG to RISCVVectorPeephole

This is split off from #71764, and moves only the vmv.v.v part of performCombineVMergeAndVOps to work on MachineInstrs.

In retrospect trying to handle PseudoVMV_V_V and PseudoVMERGE_VVM in the same function makes the code quite hard to read, so this just does it in a separate peephole.

This turns out to be simpler since for PseudoVMV_V_V we don't need to convert the Src instruction to a masked variant, and we don't need to create a fake all ones mask.

preames commented on llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:
```diff
@@ -297,6 +298,143 @@ bool RISCVVectorPeephole::convertToUnmasked(MachineInstr &MI) const {
+  // Move Src down to MI, then replace all uses of MI with it.
+  Src->moveBefore(&MI);
+
+  Src->getOperand(1).setReg(Passthru.getReg());
```
From above, we know that SrcPassThru is either NoRegister or Passthru.getReg() right?  I get needing to set it if SrcPassThru was NoRegister, but why do we need to constrain it in the other?  Should this whole block maybe be under a SrcPasThru != PassThru check?  (For readability if nothing else.)

lukel97 replied:
We don't need to constrain it if SrcPassthru = Passthru, but it should be an idempotent operation in that case anyway.

preames commented on llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:
```diff
@@ -297,6 +298,143 @@ bool RISCVVectorPeephole::convertToUnmasked(MachineInstr &MI) const {
+                           TII->getRegClass(Src->getDesc(), 1, TRI,
+                                            *Src->getParent()->getParent()));
+
+  if (MinVL->isImm())
```
Reading through this code, I'm left with a question.  If the Src instruction uses a *different* SEW than the vmv.v.v, why is it legal to reduce the VL without accounting for the different size of the elements?   I can't find that check in the DAG version of this code either.  Am I forgetting something here?

lukel97 replied:
Yikes good catch. Looks like we're miscompiling this in the DAG version too, and I don't think it's legal to fold in the mask either. Fix incoming

lukel97 replied:
This should be fixed now in this PR, but by checking the VLMAXes are the same since we don't have access to the MVTs here.

lukel97 commented on llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:
```diff
@@ -297,6 +298,143 @@ bool RISCVVectorPeephole::convertToUnmasked(MachineInstr &MI) const {
+    return false;
+
+  // Move Src down to MI, then replace all uses of MI with it.
+  Src->moveBefore(&MI);
```
We change Src's passthru to MI's passthru, and it may not be defined until after Src. So moving it to MI ensures its always available. This happens when Src's passthru is undef, which we consider equal to MI's passthru

preames replied:
This movement only needs to happen in the case where the Passthru is changing.  Can you move both the movement and the safety check to inside the if block just below?  It would make the code a bit easier to follow.  

Worth noting here is that we only need to move the Src if-and-only-if MI's psssthru is defined between the instructions.  It may be illegal to move Src in some cases that the passthru is defined *before* src. You don't need to change that in this review, it's purely a potential missed optimization.  Maybe leave a TODO though.

lukel97 replied:
I've updated it so we only perform the move if the passthru OR vl changed: The VL can get shrunk to MI's VL, in which case it could be a register defined after Src.

lukel97 commented on llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:
```diff
@@ -297,6 +298,143 @@ bool RISCVVectorPeephole::convertToUnmasked(MachineInstr &MI) const {
+    return false;
+
+  // Src needs to be a pseudo that's opted into this transform.
+  const RISCV::RISCVMaskedPseudoInfo *Info = lookupMaskedPseudoInfo(*Src);
```
> You're looking up the masked version

From what I understand RISCVMaskedPseudoInfos are only defined on the masked pseudos. This check is to keep the existing behaviour from RISCVISelDAGToDAG, but we don't actually need there to be masked pseudo available to do this fold since we're not actually converting from unmasked->masked or vice-versa.

It just happens that RISCVMaskedPseudoInfo also holds the ActiveElementsAffectResult bit.

I agree this is all confusing, I think we should move ActiveElementsAffectResult out of RISCVMaskedPseudoInfo and into TSFlags or something

lukel97 replied:
See #101123, after that we no longer need to lookup RISCVMaskedPseudos in this PR.

lukel97 committed:
Address some review comments (still more to come)

- Update comments
- Move one use check
- Rename helper to make it more clear it's only being used to check ActiveElementsAffectResult
- Only change passthru if needed

lukel97 committed:
Rebase, carry over the EEW check added in #101152

Since we don't have access to the MVTs, we can also do the same thing by checking the VLMAXs are the same (i.e. the sew/lmul ratio)

lukel97 committed:
Update to use ActiveElementsAffectResults TSFlag

lukel97 commented:
Ping

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:
```diff
@@ -65,6 +65,7 @@ class RISCVVectorPeephole : public MachineFunctionPass {
   bool convertToWholeRegister(MachineInstr &MI) const;
   bool convertToUnmasked(MachineInstr &MI) const;
   bool convertVMergeToVMv(MachineInstr &MI) const;
+  bool foldVMV_V_V(MachineInstr &MI);
```
Just a concern about the naming there. It is not consistent.

lukel97 replied:
Yeah, I noticed this as well. I think the camel case in convertVMergeToVMv came from trying to please clang-tidy. But VMV_V_V is more accurate to the actual instruction name. Do we have a preference?

wangpc-pp replied:
Hard to decide... I don't have a preference, maybe just keep it, we already have a lof of exceptions in RISCVISelLowering now.

preames reviewed (approved):
LGTM w/optional comment.

preames commented on llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:
```diff
@@ -324,6 +325,143 @@ bool RISCVVectorPeephole::convertToUnmasked(MachineInstr &MI) const {
+    return false;
+
+  bool VLChanged = !MinVL->isIdenticalTo(SrcVL);
+  bool RaisesFPExceptions = MI.getDesc().mayRaiseFPException() &&
```
Wait, shouldn't this be checking Src not MI?  MI is the vmv.v.v which will never touch fp exceptions.

If so, this implies a significant testing gap here.

lukel97 replied:
Yup, that's a typo. But it turns out we do have a test for fp exceptions on vmv.v.v and it was correctly bailing. 

Everything still happened to work because the isSafeToMove check below also checks for mayRaiseFPException, but on the correct instruction. So I've just gone ahead and deleted this bogus check here.

preames replied:
For anyone reading along, this comment is from a previous version of the code.  Luke added back the fp exceptions check when making the movement conditional (as is required for correctness).  I was initially confused by this comment, but it looks like the actual code is correct.

lukel97 committed:
Remove redundant (and incorrect) RaisesFPExceptions check

We check isSafeToMove on Src which already checks mayRaiseFPException (which in turn, also already checks for NoFPExcept

lukel97 committed:
Only move Src if needed

Because of this, we don't always call isSafeToMove so add back in the mayRaiseFPException check

preames reviewed (approved):
LGTM to the current version of the code.
